### PR TITLE
[Tests] Simple project that consumes OpenAssetIO

### PR DESCRIPTION
**Pull request by feltech**
Add a new CMake project that consumes OpenAssetIO, either as subproject or a package. CTest tests then check each component of OpenAssetIO can be imported/included/linked.

Note that Python tests in subproject mode are currently disabled, due to the complexities of marshalling Python source and binary modules. This was a pragmatic choice, since the main focus of development is toward a redistributable package, rather than use as a subproject, but will hopefully be supported at a later date.

In subproject mode `add_subdirectory` is used to bring in OpenAssetIO via a git submodule, which must be initialised externally.

In package mode `find_package` is used to bring in OpenAssetIO from an installation, which must be made available to CMake's search paths externally.

Tests are extremely minimal and do not exercise OpenAssetIO itself. The intention is just to test basic integration in a consuming project for issues with linkage etc.

### CODE REVIEW

File: src/test.core.c [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+// Include all headers.
+#include <openassetio/c/InfoDictionary.h>
+#include <openassetio/c/StringView.h>
+#include <openassetio/c/errors.h>
+#include <openassetio/c/hostApi/Manager.h>
+#include <openassetio/c/managerApi/CManagerInterface.h>
+#include <openassetio/c/managerApi/HostSession.h>
+#include <openassetio/c/managerApi/ManagerInterface.h>
+#include <openassetio/c/namespace.h>
```
Comment (foundrytom) [inline src/test.core.c:13]: [plan] Can we update the documentation in the main repo to indicate that we need to update this list when we work on the c module?

File: src/test.core.cpp [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+#include <openassetio/log/LoggerInterface.hpp>
+#include <openassetio/log/SeverityFilter.hpp>
+#include <openassetio/managerApi/Host.hpp>
+#include <openassetio/managerApi/HostSession.hpp>
+#include <openassetio/managerApi/ManagerInterface.hpp>
+#include <openassetio/managerApi/ManagerStateBase.hpp>
+#include <openassetio/trait/TraitBase.hpp>
+#include <openassetio/trait/collection.hpp>
+#include <openassetio/trait/property.hpp>
+#include <openassetio/typedefs.hpp>
```
Comment (foundrytom) [inline src/test.core.cpp:25]: [plan] Same docs note as for C

File: .editorconfig [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+indent_style = space
+insert_final_newline = true
+max_line_length = 99
+tab_width = 4
+trim_trailing_whitespace = true
+ij_continuation_indent_size = 8
+# ij_formatter_off_tag = @formatter:off
+# ij_formatter_on_tag = @formatter:on
+# ij_formatter_tags_enabled = false
+# ij_smart_tabs = false
```
Comment (foundrytom) [inline .editorconfig:16]: [question] Now were enforcing C, C++ and Python through tooling.  I'm becoming somewhat less enamoured with the army of `ij_*` entries we have here, especially unset ones.  It makes it kind of hard to grok what's actually going on, or applicable to all tools. I've been omitting them from the other new repos.  Do we think they really justify inclusion for just a single specific editor? We dont, for example list all possible generic options that we dont set, or options for other editors.

File: .editorconfig [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+ij_continuation_indent_size = 4
+# ij_prototext_keep_blank_lines_in_code = 2
+# ij_prototext_keep_indents_on_empty_lines = false
+# ij_prototext_keep_line_breaks = true
+# ij_prototext_space_after_colon = true
+# ij_prototext_space_after_comma = true
+# ij_prototext_space_before_colon = false
+# ij_prototext_space_before_comma = false
+# ij_prototext_spaces_within_braces = true
+# ij_prototext_spaces_within_brackets = false
```
Comment (foundrytom) [inline .editorconfig:98]: [consider] Removing all the file types we're highly unlikely to ever have in the project?

File: CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,179 @@
+# Copyright 2022 The Foundry Visionmongers Ltd
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.21)
+project(OpenAssetIOTest)
```
Comment (foundrytom) [inline CMakeLists.txt:5]: [consider] Do we want to update this to match the repo name?

**Review by foundrytom — CHANGES_REQUESTED** [review]
Great initiative @feltech, be awesome to have this running with CI on the main repo too.

Just a few minor notes as I'm not well placed to offer any particularly constructive CMake/C input 🙃

File: src/test.core.c (hunk 1 above)
Comment (feltech) [reply src/test.core.c:13]: Added https://github.com/OpenAssetIO/OpenAssetIO/issues/713

File: src/test.core.cpp (hunk 2 above)
Comment (feltech) [reply src/test.core.cpp:25]: See https://github.com/OpenAssetIO/OpenAssetIO-Test-CMake/pull/1#discussion_r1012044882

File: .editorconfig (hunk 3 above)
Comment (feltech) [reply .editorconfig:16]: Yeah fair enough it's quite a WOT.  I'll remove them and see if I miss them.

Comment (feltech) [reply .editorconfig]: Simplified significantly in 804ca886 - though I left a couple of the particularly useful ones (for me) in there.

File: .editorconfig (hunk 4 above)
Comment (feltech) [reply .editorconfig]: Simplified significantly in 804ca886

File: CMakeLists.txt (hunk 5 above)
Comment (feltech) [reply CMakeLists.txt]: Updated in dc370249

File: .editorconfig (hunk 3 above)
Comment (foundrytom) [reply .editorconfig]: Yeah, the guides are def a good one 👍 thanks.

File: CMakeLists.txt (hunk 5 above)
Comment (foundrytom) [reply CMakeLists.txt]: [question] Should this affect the CMAKE options and such too?

Comment (feltech) [reply CMakeLists.txt]: :thinking:

On the one hand, I could imagine one day having multiple `OpenAssetIO-Test-*` projects in a single repo, all testing different things, so we'd have to disambiguate on whatever `*` is.

But on the other hand, adding `CMAKE` to the end of the namespace could be confusing, since there are lots of built-in variables that are namespaced `CMAKE_`, so could give the impression we're overriding something built in.

I think I'd err on the side of keeping the namespace shorter.

File: README.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+the current working directory is at the root of the repository:
+
+```shell
+cmake -S . -B build
+cmake --build build
+ctest --test-dir build
+```
+
+
+### `OPENASSETIOTEST_ENABLE_SUBPROJECT=OFF` (default)
```
Comment (elliotcmorris) [inline README.md:29]: [nitpick] Double newline

File: README.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+OpenAssetIO will be discovered as a CMake package
+(`find_package(OpenAssetIO)`) and so OpenAssetIO must be pre-built and
+installed in a discoverable location.
```
Comment (elliotcmorris) [inline README.md:33]: [Consider] It's a tad unclear to me whether this description refers to the behaviour of the above option when it's the default, or when it's ON.

File: src/test.py [hunk 8]
```diff
@@ -0,0 +1,19 @@
+# Copyright 2022 The Foundry Visionmongers Ltd
+# SPDX-License-Identifier: Apache-2.0
+
+import sys
+
+
```
Comment (elliotcmorris) [inline src/test.py:6]: [nitpick] Double newline

File: .clang-tidy [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+#      `auto Foo() -> std::string { return ...; }`, we think the code is less
+#      readable in this form.
+#
+#  -modernize-return-braced-init-list: We think removing typenames and using
+#      only braced-init can hurt readability.
+#
+#  -modernize-avoid-c-arrays: We only use C arrays when they seem to be the
+#      right tool for the job, such as `char foo[] = "hello"`. In these cases,
+#      avoiding C arrays often makes the code less readable, and std::array is
+#      not a drop-in replacement because it doesn't deduce the size.
```
[the remark this answers is not in the document]
Comment (feltech) [reply .clang-tidy:22]: Yeah, most those comments are actually from google, whose style we largely copied (link at the top of the file).  It is frustrating how `std::array` can't deduce the size if you provide the type.

File: README.md (hunk 6 above)
Comment (feltech) [reply README.md]: Fixed in 23b53f95

File: README.md (hunk 7 above)
Comment (feltech) [reply README.md:33]: Hopefully clarified in 06acf102

File: src/test.py (hunk 8 above)
Comment (feltech) [reply src/test.py:6]: Fixed in 2d23a110

**Review by elliotcmorris — APPROVED** [review]

**Review by foundrytom — APPROVED** [review]
LGTMFFSAMAYL

File: .gitmodules [hunk 10]
```diff
@@ -0,0 +1,3 @@
+[submodule "OpenAssetIO"]
+	path = OpenAssetIO
+	url = https://github.com/OpenAssetIO/OpenAssetIO.git
```
Comment (foundry-markf) [inline .gitmodules:3]: Nit
Just an observation, just looking in the editor config, indent_style is space.

However, this looks like tab indentation (going by the size, and the granularity of what I can select)

File: CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,180 @@
+# Copyright 2022 The Foundry Visionmongers Ltd
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.21)
+
+project(OpenAssetIO-Test-CMake)
+
+set(CMAKE_CXX_STANDARD 17)
```
[the remark this answers is not in the document]
Comment (foundry-markf) [reply CMakeLists.txt:8]: can you tell I've been caught out by global settings before? ;)

File: CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+    # calls for dependencies must work.
+
+    add_subdirectory(OpenAssetIO)
+else ()
+    # Use OpenAssetIO as an external CMake package
+    #
+    # The `OpenAssetIOConfig.cmake` file must be discoverable. E.g.
+    # add the install directory (`dist`) to CMAKE_PREFIX_PATH.
+
+    find_package(OpenAssetIO REQUIRED)
```
Comment (foundry-markf) [inline CMakeLists.txt:52]: Should you always find OpenAssetIO by CMake config file? I assume so, as it's the modern way.

Realise that CMake will eventually find it with the correct search paths set up, but wondering whether it's better to advise this on a public test, because readers might copy-n-paste.

File: CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+            "Found Python '${Python_VERSION}' but OpenAssetIO was built against"
+            " '${OpenAssetIO_Python_VERSION}'"
+        )
+    endif ()
+
+    #-----------------------------------------------------------------------
+    # Python sources tests.
+
+    # Python sources test.
+    add_test(NAME python COMMAND ${Python_EXECUTABLE} ${PROJECT_SOURCE_DIR}/src/test.py)
```
Comment (foundry-markf) [inline CMakeLists.txt:121]: In other Python tests we've written, we've added

-v - verbose, only shown in CTest failed mode, but sometimes useful in debugging
-B - don't write .pycache files (to know you're always having to reinterpret script on re-runs of the tests
-s - ignore the user site-directory (to avoid unexpected things leaking in)

so just adding them here for consideration.

File: CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+    set_tests_properties(python PROPERTIES ENVIRONMENT PYTHONPATH=${OpenAssetIO_Python_SITELIB})
+
+    #-----------------------------------------------------------------------
+    # Python bridge test (loading Python plugin system from C++).
+
+    add_executable(test.python.bridge src/test.python.bridge.cpp)
+    add_test(python.cpp.bridge test.python.bridge)
```
Comment (foundry-markf) [inline CMakeLists.txt:128]: Nit just to say, my eyes went to 🍝 by reading the different permutations of python, bridge, cpp used in these two lines

File: CMakeLists.txt [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+    target_link_libraries(
+        test.python.bridge
+        PRIVATE
+        OpenAssetIO::openassetio-python-bridge
+        Python::Python
+    )
+
+    # Storage for list of environment variables to set before running
+    # the test.
+    set(_envvars)
```
Comment (foundry-markf) [inline CMakeLists.txt:138]: Nit: do you know you don't need to start with an empty variable?

File: CMakeLists.txt [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+            list(APPEND _envvars.path ${OpenAssetIO_BINARY_DIR_NATIVE})
+        endif ()
+        foreach (path ${Python_RUNTIME_LIBRARY_DIRS})
+            file(TO_NATIVE_PATH "${path}" path_native)
+            list(APPEND _envvars.path ${path_native})
+        endforeach ()
+        # Frustratingly, $<SEMICOLON> doesn't work here, we have to find
+        # the magic number of backslashes instead.
+        list(JOIN _envvars.path \\\; _envvars.path)
+        list(APPEND _envvars "PATH=${_envvars.path}")
```
Comment (foundry-markf) [inline CMakeLists.txt:176]: Take a peek at the USD test_package, as have had to do something similar here, but fewer backslashes.
Had to distinguish between semi-colons for CMake list purposes, and semi-colons for Windows PATH separators.

File: README.md [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+cmake --build build
+ctest --test-dir build
+```
+
+### `OPENASSETIOTEST_ENABLE_SUBPROJECT=OFF` (default)
+
+When `OPENASSETIOTEST_ENABLE_SUBPROJECT=OFF` (which is the default if
+unspecified), OpenAssetIO will be discovered as a CMake package
+(`find_package(OpenAssetIO)`) and so OpenAssetIO must be pre-built and
+installed in a discoverable location.
```
Comment (foundry-markf) [inline README.md:33]: Worth being explicit here, and say, add the installed location to CMAKE_PREFIX_PATH?

File: src/test.core.c (hunk 1 above)
Comment (foundry-markf) [inline src/test.core.c:13]: Just an observation from the outside world... you've got almost as many lines of #includes as there are in the body of main.

Are there thoughts about a utility header "#include <openassetio/c/openassetio.h>" for example that brings them all in?

File: src/test.core.c [hunk 18]
```diff
@@ -0,0 +1,30 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2022 The Foundry Visionmongers Ltd
+#include <stdio.h>
```
Comment (foundry-markf) [inline src/test.core.c:3]: Failing to find a good reference (some people's personal opinions on SO and the like), but I would always put standard library #includes _last_.

Reason? If any of the project headers need stdio, but don't #include it, then this is masking another consume not doing the initial #include for stdio.h.

File: src/test.core.cpp (hunk 2 above)
Comment (foundry-markf) [inline src/test.core.cpp:25]: Blimey!

All needed for two lines of C++ code?

File: src/test.py [hunk 19]
```diff
@@ -0,0 +1,18 @@
+# Copyright 2022 The Foundry Visionmongers Ltd
+# SPDX-License-Identifier: Apache-2.0
+
+import sys
+
+try:
+    import openassetio
+    from openassetio.hostApi import ManagerFactory
+except ImportError as exc:
+    print("Error importing openassetio package: {}".format(exc))
```
Comment (foundry-markf) [inline src/test.py:10]: :D This looks suspiciously formatted like some of our existing tests ;) :D

Q: Is OpenAssetIO supported on Python 2 at all?

I know it probably shouldn't be with modern development, but, some studios are still using it.

`print` is what I'm thinking about for Py2.

File: src/test.python.bridge.cpp [hunk 20]
```diff
@@ -0,0 +1,28 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2022 The Foundry Visionmongers Ltd
+
+#include <memory>
+
+#include <Python.h>
```
Comment (foundry-markf) [inline src/test.python.bridge.cpp:6]: Would move these to the end of the #include list, and swap their order.

File: CMakeLists.txt (hunk 12 above)
Comment (feltech) [reply CMakeLists.txt:52]: Unfortunately not for the case of public ConanCenter packages, according to [their docs](https://github.com/conan-io/conan-center-index/blob/master/docs/packaging_policy.md#package)

> CMake config files must be removed (they will be generated for consumers by cmake_find_package, cmake_find_package_multi or CMakeDeps generators).

The (eventual) public conanfile will have to copy the `OpenAssetIOConfig` logic into a generated `FindOpenAssetIO` module :disappointed:

File: src/test.python.bridge.cpp [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+
+#include <openassetio/log/LoggerInterface.hpp>
```
Comment (foundry-markf) [inline src/test.python.bridge.cpp:8]: Could start a whole flame war about quotes vs angle brackets.

But, the point I'd make, is that it depends on _where_ you install libraries as to which is appropriate, especially on compilers that distinguish include search path command line arguments between user and system paths.

In Conan land, these files will _never_ be in a system location.
VCPKG they won't be either (although in their own /usr like structure).
rez, I don't know.

All I'm saying is that "it depends" whether this is considered a system library or not.

File: src/test.python.bridge.cpp [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+  void log(Severity severity, const openassetio::Str& message) override {}
+};
+}  // namespace
+
+int main() {
+  Py_Initialize();
+  auto logger = std::make_shared<DummyLoggerInterface>();
+  {
+    const auto factory =
+        openassetio::python::hostApi::createPythonPluginSystemManagerImplementationFactory(logger);
```
Comment (foundry-markf) [inline src/test.python.bridge.cpp:23]: A mouthful.... just saying. You'll be competing with Katana soon.

File: CMakeLists.txt (hunk 14 above)
Comment (feltech) [reply CMakeLists.txt:128]: :grin: Yeah it's tricky, I didn't want to use the word "test" in what is already obviously a test, and didn't want to use the word "cpp" in what is obviously a `.cpp` file, but then also wanted to capture those things in the source file and test target name, respectively.

File: src/test.python.bridge.cpp [hunk 23]
```diff
@@ -0,0 +1,28 @@
...
+#include <openassetio/python/hostApi.hpp>
+
+namespace {
+struct DummyLoggerInterface : openassetio::log::LoggerInterface {
+  void log(Severity severity, const openassetio::Str& message) override {}
+};
+}  // namespace
+
+int main() {
+  Py_Initialize();
```
Comment (foundry-markf) [inline src/test.python.bridge.cpp:19]: I've just checked what I've done with embedding Python interpreters before, and I've always set Py_SetProgramName before Py_Initialize.

Also called `Py_SetPath` _after_, but this was a GUI application, so wasn't appropriate to get that from an environment variable like in here. But just saying it for completeness.

File: CMakeLists.txt (hunk 15 above)
Comment (feltech) [reply CMakeLists.txt:138]: I thought this would be best practice, in case at some point in the future `_envvars` was used above, so this would "reset" it back to blank.  I did toy with the idea of a really unique name like `python.cpp.bridge.envars` to guard against that instead, but thought it a bit wordy and obfuscating.

File: CMakeLists.txt (hunk 16 above)
Comment (feltech) [reply CMakeLists.txt:176]: I believe the number of required backlashes corresponds to the number of times the variable is substituted.  Here there is the JOIN then the APPEND then the final ENVIRONMENT substitution.  I'm not sure at what point each backslash is collapsed, since it was more-or-less a trial-and-error thing.

File: CMakeLists.txt (hunk 12 above)
Comment (foundry-markf) [reply CMakeLists.txt:52]: Well, this is interesting, because I was on the Conan 2 hackathon call this week, and it was brought up that packages can provide their own CMake config files, and you can describe in receipe when they do. I even asked a question on it!

File: CMakeLists.txt (hunk 14 above)
Comment (foundry-markf) [reply CMakeLists.txt:128]: Might be easier if python bridge and cpp were always in the same order?

File: CMakeLists.txt (hunk 15 above)
Comment (foundry-markf) [reply CMakeLists.txt:138]: > so this would "reset" it back to blank.

True

No harm in leaving, just wanted to make sure you knew that it wasn't strictly necessary.

File: README.md (hunk 17 above)
Comment (feltech) [reply README.md:33]: I will add that as an example, but of course there are many alternatives to `CMAKE_PREFIX_PATH`.

Comment (foundry-markf) [reply README.md:33]: Well, that was also when I was thinking only config files were being used.

How about saying "in a discoverable location matching the documented find_package search options" (or similar)?

File: src/test.core.c (hunk 1 above)
Comment (feltech) [reply src/test.core.c:13]: This is specifically to test that all expected headers exist and can compile (in this case in-particular with a C compiler). We wouldn't expect users to include all the headers in normal usage.

Having said that, it would be handy to ease maintenance of this repo if OpenAssetIO maintained an uber header instead :thinking:.

File: src/test.core.c (hunk 18 above)
Comment (feltech) [reply src/test.core.c:3]: That's a really compelling argument, I hadn't considered that. 

In general our coding style has tried to keep C++ and Python as similar as possible, and Python's PEP8 says system headers (imports) come first.

Handily, this also somewhat matches cpplint (which apparently cannot be configured otherwise), which is based on [Google's style guide](https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes), which we decided early on to follow (for the most part).  I.e.

> Include headers in the following order: Related header, C system headers, C++ standard library headers, other libraries' headers, your project's headers.

File: src/test.core.c (hunk 1 above)
Comment (foundry-markf) [reply src/test.core.c:13]: > Having said that, it would be handy to ease maintenance of this repo if OpenAssetIO maintained an uber header instead

Would users also find it useful? is there any visibility on whether they're regularly #include'ing the majority of the exposed public headers?


Q: Do you expect all headers to each be standalone?

In a previous company, we used to have tests that autogenerated a c file per header, which just had a single #include for its corresponding header, and was compiled (not linked, just compiled).

It then confirmed whether that header had any missing #includes itself, or forward declarations, to avoid consumers needing to know about them.

File: src/test.core.c (hunk 18 above)
Comment (foundry-markf) [reply src/test.core.c:3]: Yep, agree it's inverted to Python, as I just got it confused on another review I'm doing today, and needed to go and look up the PEP! :P

File: src/test.core.cpp (hunk 2 above)
Comment (feltech) [reply src/test.core.cpp:25]: :grin: As mentioned in https://github.com/OpenAssetIO/OpenAssetIO-Test-CMake/pull/1#discussion_r1013846233 this is just used to check that headers exist and can be compiled.

I'll update the code comment to make this clear.

File: src/test.py (hunk 19 above)
Comment (feltech) [reply src/test.py:10]: Thankfully, no Python 2 for us!

Happy to improve the formatting if you have a suggestion?

File: src/test.core.c (hunk 18 above)
Comment (foundry-markf) [reply src/test.core.c:3]: Another way to ask the question is two-fold:
* how many system headers do OpenAssetIO headers depend upon? (if none, then moot point)
* how annoyed do you think users would be if consumer code failed to compile because an OpenAssetIO header is erroring with 'can't find <xyz> definition'? - of course, these are developers as users, so they'd solve it, but you might get a bug report.

File: src/test.py (hunk 19 above)
Comment (foundry-markf) [reply src/test.py:10]: Formatting and structure is fine IMO (biased, as I created the first instance!)

It was merely the use of print that made me ask the question about Py2.

File: src/test.python.bridge.cpp (hunk 21 above)
Comment (feltech) [reply src/test.python.bridge.cpp:8]: My take on angle brackets vs. quotes is to reserve quotes for relative paths (since IIRC angle brackets don't work at all for that, at least on GCC).  There's a ticket (https://github.com/OpenAssetIO/OpenAssetIO/issues/424) over on the OpenAssetIO repo related to this.

Although Conan does not install into a system location, the include directories are typically added to the `gcc` command line using `-isystem`, so are at least treated in a similar way to system headers.

File: src/test.python.bridge.cpp (hunk 22 above)
Comment (feltech) [reply src/test.python.bridge.cpp:23]: :laughing: Yeah we had great fun coming up with this name. Hardest problem in software engineering, naming things.

File: src/test.python.bridge.cpp (hunk 21 above)
Comment (foundry-markf) [reply src/test.python.bridge.cpp:8]: Happy there's a ticket on it.
Don't envy the task of making everyone happy on a decision though. Sorry.

File: CMakeLists.txt (hunk 14 above)
Comment (feltech) [reply CMakeLists.txt:128]: Yeah OK if that reads easier.  I had in my mind "the Python-to-C++ bridge" so `python.cpp.bridge` kinda worked, but `python.bridge.cpp` works too.

File: src/test.core.c (hunk 18 above)
Comment (feltech) [reply src/test.core.c:3]: I think it may be out of scope to change our style guide in this PR, though I'll raise it with the team.  The main argument in this specific case, i.e. ensuring no implicit dependencies between headers, would be solved by your other excellent suggestion in https://github.com/OpenAssetIO/OpenAssetIO-Test-CMake/pull/1#discussion_r1013856218 (to have a separate compilation unit per header).

File: src/test.python.bridge.cpp (hunk 23 above)
Comment (feltech) [reply src/test.python.bridge.cpp:19]: I had a quick play with `Py_SetProgramName` in Windows and Linux.  

Firstly, its 4 extra lines of code (`wchar_t` conversion) - good 'ole C.

In Windows I didn't see any real change in behaviour.

In Linux, on the other hand, if I point `Py_SetProgramName` to the absolute path of the Python interpreter in the Conan package (i.e. the same package that provides the linked `libpython`), then I can remove the `PYTHONHOME` environment variable, since `Py_Initialize` computes paths based on the program name.

On `Py_SetPath` [the docs](https://docs.python.org/3.7/c-api/init.html#c.Py_SetPath) say

>  This is useful if Python is embedded by an application that has full knowledge of the location of all modules

So setting these in the source code would allow us to remove the `PYTHONHOME`  and  `PYTHONPATH` environment variables, but at the cost of passing their values along as compile definitions instead, essentially hardcoding them in the executable.

I don't think adding either of these makes the executable any more "correct", yet would add a bit more complexity to something that already works as is, so I'm of a mind to leave it be, if that's OK?

File: .gitmodules (hunk 10 above)
Comment (feltech) [reply .gitmodules:3]: Ah just realised, this is an autogenerated file, so I think OK to leave it as it is.

Comment (foundry-markf) [reply .gitmodules:3]: Autogenerated by what?
I thought I've had to craft these by hand before.

File: CMakeLists.txt (hunk 16 above)
Comment (foundry-markf) [reply CMakeLists.txt:176]: Sure, it was trial and error for me on USD too. Just wanted to point out that it had been tackled elsewhere if you want to take a look.

File: src/test.core.c (hunk 1 above)
Comment (foundry-markf) [reply src/test.core.c:13]: Of course. It's out of scope for this PR.

File: .gitmodules (hunk 10 above)
Comment (feltech) [reply .gitmodules:3]: `git submodule add` IIRC

File: CMakeLists.txt (hunk 11 above)
[the remark this answers is not in the document]
Comment (feltech) [reply CMakeLists.txt]: Set per-target rather than globally in f4ce8f82

File: CMakeLists.txt (hunk 13 above)
Comment (feltech) [reply CMakeLists.txt]: I found that `-v` was an insane amount of logging!  So I went with just `-Bs` in 8c253c09

File: README.md (hunk 17 above)
Comment (feltech) [reply README.md]: Does this read a bit better: 92b5bd49 ?

File: src/test.core.c (hunk 1 above)
Comment (feltech) [reply src/test.core.c]: Added https://github.com/OpenAssetIO/OpenAssetIO-Test-CMake/issues/2

File: src/test.core.cpp (hunk 2 above)
Comment (feltech) [reply src/test.core.cpp:26]: Updated code comment in d5fdb87a

File: src/test.python.bridge.cpp (hunk 20 above)
Comment (feltech) [reply src/test.python.bridge.cpp:6]: Discussed further in https://github.com/OpenAssetIO/OpenAssetIO-Test-CMake/pull/1#discussion_r1013815282

File: CMakeLists.txt (hunk 12 above)
Comment (feltech) [reply CMakeLists.txt:51]: Maybe this is a ConanCenter-specific policy? Or perhaps the policy will change with Conan 2?

File: CMakeLists.txt (hunk 13 above)
Comment (foundry-markf) [reply CMakeLists.txt]: But you only see that on a failure?

File: src/test.python.bridge.cpp (hunk 23 above)
Comment (foundry-markf) [reply src/test.python.bridge.cpp:19]: Yeah that's fine if it doesn't benefit you, I just wanted to mention those other parts that I've needed to use before, so good you've had a look over them. Thanks!

File: .gitmodules (hunk 10 above)
Comment (foundry-markf) [reply .gitmodules:3]: TIL that there are tabs in other .gitmodules I've just looked at in the ecosystem! So... moot point. Nothing to see here.

File: CMakeLists.txt (hunk 12 above)
Comment (foundry-markf) [reply CMakeLists.txt:51]: My impression was that it's in the latest Conan 1.x too. But maybe it'll change. Anyway, you need to be compatible with their policies as they are to get through a PR to Conan Center.

File: CMakeLists.txt (hunk 13 above)
Comment (feltech) [reply CMakeLists.txt]: I guess that's true (I always run with `--extra-verbose` locally), but even then I'd worry about that much output obfuscating the actual error.  

I recall when I discovered the need to set `PYTHONHOME` that the error spat out by Python was quite informative and verbose.

Do you really think `-v` is more useful than noise?

Comment (foundry-markf) [reply CMakeLists.txt]: It _has_ been useful. But if you're more comfortable without it, I won't press.

We're mostly debugging when thirdparty libraries aren't behaving, as opposed to our own code in the equivalent tests, hence more logging is useful.

Comment (feltech) [reply CMakeLists.txt]: OK, happy to defer to experience - added `-v` in 1d39c428

**Comment by foundry-markf** [comment]
I'm good with this.
Would I be right in thinking this recipe pipeline covers this latest code?
https://gitlab.thefoundry.co.uk/libraries/conan/recipes/-/pipelines/54924

**Review by foundry-markf — APPROVED** [review]
LGTM
(and confirmed on the recipes.git pipeline)
